cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -89,21 +81,71 @@ type CertManagerClient interface {
 
 // certManagerClient implements CertManagerClient .
 type certManagerClient struct {
-	configClient        config.Client
-	proxy               Proxy
-	pollImmediateWaiter PollImmediateWaiter
+	configClient                       config.Client
+	proxy                              Proxy
+	pollImmediateWaiter                PollImmediateWaiter
+	embeddedCertManagerManifestVersion string
+	embeddedCertManagerManifestHash    string
 }
 
 // Ensure certManagerClient implements the CertManagerClient interface.
 var _ CertManagerClient = &certManagerClient{}
 
+func (cm *certManagerClient) setManifestHash() error {
+	yamlData, err := manifests.Asset(embeddedCertManagerManifestPath)
+	if err != nil {
+		return err
+	}
+	cm.embeddedCertManagerManifestHash = fmt.Sprintf("%x", sha256.Sum256(yamlData))
+	return nil
+}
+
+func (cm *certManagerClient) setManifestVersion() error {
+	// Gets the cert-manager objects from the embedded assets.
+	objs, err := cm.getManifestObjs()
+	if err != nil {
+		return err
+	}
+	found := false
+
+	for i := range objs {
+		o := objs[i]
+		if o.GetKind() == "CustomResourceDefinition" {
+			labels := o.GetLabels()
+			version, ok := labels[certmanagerVersionLabel]
+			if ok {
+				s := strings.Split(version, "-")
+				cm.embeddedCertManagerManifestVersion = s[2]
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		return errors.Errorf("Failed to detect cert-manager version by searching for label %s in all CRDs", certmanagerVersionLabel)
+	}
+	return nil
+}
+
 // newCertMangerClient returns a certManagerClient.
 func newCertMangerClient(configClient config.Client, proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *certManagerClient {
-	return &certManagerClient{
+	log := logf.Log
+
+	cm := &certManagerClient{
 		configClient:        configClient,
 		proxy:               proxy,
 		pollImmediateWaiter: pollImmediateWaiter,
 	}
+	err := cm.setManifestVersion()
+	if err != nil {
+		log.Info("failed to set Manifest Version")
+	}
+
+	err = cm.setManifestHash()
+	if err != nil {
+		log.Info("failed to set Manifest Hash")
```

Comment:
Similar to above comment.
Except if we happen to have a `cm.embeddedCertManagerManifestVersion` set but then have an error in `cm.setManifestHash()`, we again have an empty value for `cm.embeddedCertManagerManifestHash`. 

So in `cm.shouldUpgrade()`, we'd be comparing the actual hash stored in previous versions with an empty string and then forcing an update.
```golang
if !ok || objHash != cm.embeddedCertManagerManifestHash {
``` 

https://github.com/kubernetes-sigs/cluster-api/pull/3707/files#diff-4c9090c205e1c8908f4973e2fc4f334ab689c3446d6a207fec2d930e0150b3ffR337

Reply:
thanks for point out
this logic seems like: 
1) if annotation not exist, force upgrade
2) if the annotation exist, then check whether hash is same to current version's hash
if not same, force upgrade,
in our case, if hash is `''`, so it will be force upgrade.. thus current logic seems ok?

Reply:
So for `2`, Is the correct to have an empty hash in the first place? 

This is my thought process. 
If `embeddedCertManagerManifestHash` is empty and we see that the currently installed cert-manager has a hash, we go on to upgrade it, but then as part of the upgrade in `EnsureLatestVersion` we do an `cm.install()` which then proceeds to add the `certmanagerHashAnnotation` to the objects which now contains an empty string. So we've effectively removed the hash annotation in this case which will affect future upgrades.

My thought is that because we embed cert-manager within clusterctl, it should always have a hash and version for the cert-manager. These fields should not be empty. If they are empty, then something else is gone wrong and we need to fail and let the user know.